sdk/core/Azure.Core/src/HttpMessage.cs

```diff
@@ -80,19 +81,46 @@ public Response Response
         /// </summary>
         public TimeSpan? NetworkTimeout { get; set; }
 
-        internal void AddPolicies(RequestContext? context)
+        internal void ApplyRequestContext(RequestContext? context)
         {
-            if (context == null || context.Policies == null || context.Policies.Count == 0)
+            if (context == null)
             {
                 return;
             }
 
-            Policies ??= new(context.Policies.Count);
-            Policies.AddRange(context.Policies);
+            if (context.Policies != null &&  context.Policies.Count > 0)
+            {
+                Policies ??= new(context.Policies.Count);
+                Policies.AddRange(context.Policies);
+            }
+
+            _errorCodes = context.ErrorCodes;
+            _nonErrorCodes = context.NonErrorCodes;
         }
 
         internal List<(HttpPipelinePosition Position, HttpPipelinePolicy Policy)>? Policies { get; set; }
 
+        private int[]? _errorCodes { get; set; }
+        private int[]? _nonErrorCodes { get; set; }
```

Comment:
I think this will not usually have more than one or two entries.  I would also worry more about allocating a full object instead of just a few bytes for the array, but I haven't tested it explicitly.

Reply:
I ran a benchmark out of curiosity - if we include the allocations to create the hash, it's worse all up by 4x to use a HashSet. If we measure just the lookup, we are 3x faster with a HashSet. So, unless we had a model where the error/non-error codes were re-used across messages, using a HashSet is worse.